src/IO/RemoteReadBufferCache.cpp

```diff
@@ -0,0 +1,585 @@
+
+    /// Position is unchanged.
+    if (new_pos + (working_buffer.end() - pos) == pos_in_file)
+        return new_pos;
+
+    if (new_pos <= pos_in_file && new_pos >= pos_in_file - static_cast<off_t>(working_buffer.size()))
+    {
+        /// Position is still inside buffer.
+        pos = working_buffer.begin() + (new_pos - (pos_in_file - working_buffer.size()));
+        return new_pos;
+    }
+
+    pos = working_buffer.end();
+    auto ret_off = file_reader->seek(new_pos);
+    if (ret_off == -1)
+        throw Exception(
+            "seek file failed. " + std::to_string(pos_in_file) + "->" + std::to_string(new_pos) + "@" + std::to_string(file_reader->size())
+                + "," + std::to_string(whence) + "," + file_reader->getPath(),
+            ErrorCodes::BAD_ARGUMENTS);
+    return ret_off;
+}
+
+off_t RemoteReadBuffer::getPosition()
+{
+    return file_reader->getOffset() - available();
+}
+
+RemoteReadBufferCache::RemoteReadBufferCache() = default;
+
+RemoteReadBufferCache::~RemoteReadBufferCache()
+{
+    threadPool->wait();
+}
+
+RemoteReadBufferCache & RemoteReadBufferCache::instance()
+{
+    static RemoteReadBufferCache instance;
+    return instance;
+}
+
+void RemoteReadBufferCache::recoverCachedFilesMeta(
+        const std::filesystem::path & current_path,
+        size_t current_depth,
+        size_t max_depth,
+        std::function<void(RemoteCacheController *)> const & finish_callback)
+{
+    if (current_depth >= max_depth)
+    {
+        for (auto const & dir : std::filesystem::directory_iterator{current_path})
+        {
+            std::string path = dir.path();
+            auto cache_controller = RemoteCacheController::recover(path, finish_callback);
+            if (!cache_controller)
+                continue;
+            auto &cell = caches[path];
+            cell.cache_controller = cache_controller;
+            cell.key_iterator = keys.insert(keys.end(), path);
+        }
+        return;
+    }
+
+    for (auto const &dir : std::filesystem::directory_iterator{current_path})
+    {
+        recoverCachedFilesMeta(dir.path(), current_depth + 1, max_depth, finish_callback);
+    }
+
+}
+
+void RemoteReadBufferCache::initOnce(const std::filesystem::path & dir,
+        size_t limit_size_,
+        size_t bytes_read_before_flush_,
+        size_t max_threads)
+{
+    LOG_TRACE(log, "init local cache. path: {}, limit {}", dir.string(), limit_size_);
+    local_path_prefix = dir;
+    limit_size = limit_size_;
+    local_cache_bytes_read_before_flush = bytes_read_before_flush_;
+    threadPool = std::make_shared<FreeThreadPool>(max_threads, 1000, 1000, false);
+
+    // scan local disk dir and recover the cache metas
+    std::filesystem::path root_dir(local_path_prefix);
+    if (!std::filesystem::exists(root_dir))
+    {
+        LOG_INFO(log, "{} not exists. this cache will be disable", local_path_prefix);
+        return;
```

Comment:
Yes. You are right.

Reply:
Well, may be we should do that in code and not the user (at least if the directory does not exist yet)? I think it is better to have a setting which enables / disables local caching rather than counting on user must create a directory himself.

Reply:
Ok,  I will add a switch in settings.

Reply:
done